File: acmOtherMaterials/Ulm Local Contest/2004/solution/code.c
/* Problem   Code
** Algorithm Euler Tour (Depth First Search)
** Runtime   O(10^(n+1))
** Author    Walter Guttmann
** Date      2000.07.13
*/

#include <assert.h>
#include <stdio.h>

int edge[100000][10], nstates;
char out[1048576], *outp;

void dfs(int q)
{
  int i;
  /* We are in state q and press the key i. */
  for (i=0 ; i<10 ; i++)
    if (edge[q][i])
    {
      edge[q][i] = 0;
      dfs((q * 10 + i) % nstates);
      *--outp = '0' + i;
    }
}

int main()
{
  FILE *in = fopen("code.in", "r");
  while (1)
  {
    int i, n, q;
    fscanf(in, " %d ", &n);
    if (n == 0) break;
    assert(1 <= n && n <= 6);
    /* There are 10^(n-1) states. */
    nstates = 1;
    for (i=1 ; i<n ; i++)
      nstates *= 10;
    /* From every state, by pressing a key, one of 10 states is reachable. */
    for (q=0 ; q<nstates ; q++)
      for (i=0 ; i<10 ; i++)
        edge[q][i] = 1;
    /* We always start with pressing the 0 key n-1 times. */
    for (i=1 ; i<n ; i++)
      printf("0");
    /* Produce an Euler tour in reverse order and output it. */
    *(outp = out + nstates*10) = 0;
    dfs(0);
    assert(outp == out);
    printf("%s\n", out);
  }
  return 0;
}

File: acmOtherMaterials/Ulm Local Contest/2004/solution/code.java
// Problem   Code
// Algorithm Euler Tour (Iterative Depth First Search)
// Runtime   O(10^(n+1))
// Author    Walter Guttmann
// Date      2000.07.13

import java.io.*;

public class code
{
  public static void main(String[] arg) throws Exception
  {
    StreamTokenizer st = new StreamTokenizer(new FileReader("code.in"));
    while (true)
    {
      st.nextToken();
      int n = (int) st.nval;
      if (n == 0) break;
      // There are 10^(n-1) states.
      int nstates = 1;
      for (int k=1 ; k<n ; k++)
        nstates *= 10;
      // From every state, by pressing a key, one of 10 states is reachable.
      boolean[][] edge = new boolean[nstates][10];
      for (int q=0 ; q<nstates ; q++)
        for (int i=0 ; i<10 ; i++)
          edge[q][i] = true;

      // Produce an Euler tour in reverse order.
      int nout = nstates*10;
      char[] out = new char[nout];

      int nstack = 0;
      int stack_q[] = new int[nout+1];
      int stack_i[] = new int[nout+1];
      stack_q[0] = 0;
      stack_i[0] = 0;

      while (true)
      {
        int q = stack_q[nstack];
        int i = stack_i[nstack];
        if (i < 10)
        {
          stack_i[nstack]++;
          if (edge[q][i])
          {
            edge[q][i] = false;
            ++nstack;
            stack_q[nstack] = (q * 10 + i) % nstates;
            stack_i[nstack] = 0;
          }
        }
        else
        {
          if (nstack == 0)
            break;
          --nstack;
          out[--nout] = (char) ((int) '0' + stack_i[nstack] - 1);
        }
      }

      CharArrayWriter caw = new CharArrayWriter();
      // We always start with pressing the 0 key n-1 times.
      for (int k=1 ; k<n ; k++)
        caw.write('0');
      caw.write(out, 0, nstates*10);
      System.out.println(caw.toString());
    }
  }
}

